Include external layers into `querybypolygon` search results (ref: `addlayers` map control )

volterra79 opened the pull request:
Closes: #355 

Steps:

- Add external layer using **Add Layer** map control

![Screenshot from 2023-02-21 15-27-12](https://user-images.githubusercontent.com/1051694/220372296-4de83394-2675-4b01-9787-83942d23ccfd.png)


![Screenshot from 2023-02-21 15-28-54](https://user-images.githubusercontent.com/1051694/220372618-6c7b6fa8-7baa-4bf0-a9d9-c14168cf001c.png)

- Use **Query By Polygon** map control 

![Screenshot from 2023-02-21 15-31-57](https://user-images.githubusercontent.com/1051694/220373356-f1fd2907-170f-49c5-bce9-a5fc6fb21a1d.png)

we can see below that feature from added layer is show on results

![Screenshot from 2023-02-21 15-32-51](https://user-images.githubusercontent.com/1051694/220373747-0986697e-a0f9-4185-8f60-149f3896ee2d.png)

volterra79 committed:
Add external layer listen selected property to enable or not the control

volterra79 committed:
Handle case of add layer external layer selected property

volterra79 committed:
Add handle of selected property on external layer (vector)

volterra79 committed:
add selected property to external layer

volterra79 committed:
add intersects and within methods to check if two OL geometries are intersected o one geometry is within

volterra79 committed:
Case of layersFeaturesBoxes[boxid] is undefined

volterra79 committed:
Handle external vector layer in case of query type. It add feature based on selected layer external

volterra79 committed:
Merge branch 'dev' into v3.6_feature-add-external-layer-to-querybypolygon

# Conflicts:
#	src/components/CatalogTristateTree.vue

volterra79 committed:
change order of undefined typeof check

volterra79 committed:
removed console.log

volterra79 committed:
Need to wait promise resolution of setter setSelectedExternalLayer catalogService

volterra79 committed:
Merge branch 'dev' into v3.6_feature-add-external-layer-to-querybypolygon

volterra79 committed:
Use feature.getId() method to get external fid info of a query by polygon

Raruto commented:
I'm not sure if it's a bug or somehow related only to this pull, but testing it I noticed the following:

- external layer: [example.kmz.zip](https://github.com/g3w-suite/g3w-client/files/10813013/example.kmz.zip) - you need to download it rename to `.kmz` because github.com doesn't allow uploading of kmz files..
- project URL: http://localhost:8000/it/map/buildings/qdjango/10/
- style URL: http://localhost:8000/it/map/buildings/qdjango/10/#m_ylw-pushpin

By clicking on the `Style URL` link you will be redirected to above location.

![image](https://user-images.githubusercontent.com/9614886/220893561-7dbb846c-8b81-4829-ab4c-f3539631e86d.png)

Raruto committed:
update some jsdoc

Raruto committed:
jsdoc

volterra79 commented:
> I'm not sure if it's a bug or somehow related only to this pull, but testing it I noticed the following:
> 
> * external layer: [example.kmz.zip](https://github.com/g3w-suite/g3w-client/files/10813013/example.kmz.zip) - you need to download it rename to `.kmz` because github.com doesn't allow uploading of kmz files..
> * project URL: http://localhost:8000/it/map/buildings/qdjango/10/
> * style URL: http://localhost:8000/it/map/buildings/qdjango/10/#m_ylw-pushpin
> 
> By clicking on the `Style URL` link you will be redirected to above location.
> 
> ![image](https://user-images.githubusercontent.com/9614886/220893561-7dbb846c-8b81-4829-ab4c-f3539631e86d.png)

@Raruto i test it an i let you know

Raruto committed:
add new function `QueryByPolygonControl::_handleExternalLayers()`

Raruto commented on src/app/g3w-ol/controls/querybypolygoncontrol.js:
```diff
@@ -1,15 +1,28 @@
+   * 
+   * @since 3.8.0
+   */
+  this.externalLayer = [];
```
Perhaps the plurar for `this.externalLayers` could be more appropriate?

volterra79 replied:
@Raruto sure

volterra79 replied:
@Raruto it seems that is an issue related to ol.format.KML

```xml
<?xml version="1.0" encoding="UTF-8"?>
<kml xmlns="http://www.opengis.net/kml/2.2" xmlns:gx="http://www.google.com/kml/ext/2.2" xmlns:kml="http://www.opengis.net/kml/2.2" xmlns:atom="http://www.w3.org/2005/Atom">
<Document>
	<name>example.kmz</name>
	<StyleMap id="m_ylw-pushpin">
		<Pair>
			<key>normal</key>
			<styleUrl>#s_ylw-pushpin</styleUrl>
		</Pair>
		<Pair>
			<key>highlight</key>
			<styleUrl>#s_ylw-pushpin_hl</styleUrl>
		</Pair>
	</StyleMap>
	<Style id="s_ylw-pushpin">
		<IconStyle>
			<scale>1.1</scale>
			<Icon>
				<href>http://maps.google.com/mapfiles/kml/pushpin/ylw-pushpin.png</href>
			</Icon>
			<hotSpot x="20" y="2" xunits="pixels" yunits="pixels"/>
		</IconStyle>
	</Style>
	<Style id="s_ylw-pushpin_hl">
		<IconStyle>
			<scale>1.3</scale>
			<Icon>
				<href>http://maps.google.com/mapfiles/kml/pushpin/ylw-pushpin.png</href>
			</Icon>
			<hotSpot x="20" y="2" xunits="pixels" yunits="pixels"/>
		</IconStyle>
	</Style>
	<Placemark>
		<name>example</name>
		<styleUrl>#m_ylw-pushpin</styleUrl>
		<Polygon>
			<tessellate>1</tessellate>
			<outerBoundaryIs>
				<LinearRing>
					<coordinates>
						11.24608850202978,43.79212956957394,0 11.24541318716863,43.79061166470144,0 11.24838553990347,43.78985230128217,0 11.24884032778454,43.79167008097611,0 11.24826933207685,43.79230733776345,0 11.24608850202978,43.79212956957394,0 
					</coordinates>
				</LinearRing>
			</outerBoundaryIs>
		</Polygon>
	</Placemark>
</Document>
</kml>
```
Above is the kml extract from example.kmz. It get styleUrl as feature property

Raruto committed:
jsdoc

volterra79 committed:
Rename externalLayer to externalLayers

volterra79 committed:
Referred to https://github.com/g3w-suite/g3w-client/pull/356#issuecomment-1441609067

Raruto commented on src/app/core/utils/geo.js:
```diff
@@ -736,6 +736,9 @@ const geoutils = {
         featureProjection: mapCrs || epsg
       });
       if (features.length) {
+        if (format instanceof ol.format.KML) {
+          features.forEach(feature => feature.unset('styleUrl'));
+        }
```
Whoa what a roughness! 🦆 🤭

Raruto commented:
> @Raruto it seems that is an issue related to ol.format.KML
> 
> ```
> <?xml version="1.0" encoding="UTF-8"?>
> <kml xmlns="http://www.opengis.net/kml/2.2" xmlns:gx="http://www.google.com/kml/ext/2.2" xmlns:kml="http://www.opengis.net/kml/2.2" xmlns:atom="http://www.w3.org/2005/Atom">
> <Document>
> 	<name>example.kmz</name>
> 	<StyleMap id="m_ylw-pushpin">
> 		<Pair>
> 			<key>normal</key>
> 			<styleUrl>#s_ylw-pushpin</styleUrl>
> 		</Pair>
> 		<Pair>
> 			<key>highlight</key>
> 			<styleUrl>#s_ylw-pushpin_hl</styleUrl>
> 		</Pair>
> 	</StyleMap>
> 	<Style id="s_ylw-pushpin">
> 		<IconStyle>
> 			<scale>1.1</scale>
> 			<Icon>
> 				<href>http://maps.google.com/mapfiles/kml/pushpin/ylw-pushpin.png</href>
> 			</Icon>
> 			<hotSpot x="20" y="2" xunits="pixels" yunits="pixels"/>
> 		</IconStyle>
> 	</Style>
> 	<Style id="s_ylw-pushpin_hl">
> 		<IconStyle>
> 			<scale>1.3</scale>
> 			<Icon>
> 				<href>http://maps.google.com/mapfiles/kml/pushpin/ylw-pushpin.png</href>
> 			</Icon>
> 			<hotSpot x="20" y="2" xunits="pixels" yunits="pixels"/>
> 		</IconStyle>
> 	</Style>
> 	<Placemark>
> 		<name>example</name>
> 		<styleUrl>#m_ylw-pushpin</styleUrl>
> 		<Polygon>
> 			<tessellate>1</tessellate>
> 			<outerBoundaryIs>
> 				<LinearRing>
> 					<coordinates>
> 						11.24608850202978,43.79212956957394,0 11.24541318716863,43.79061166470144,0 11.24838553990347,43.78985230128217,0 11.24884032778454,43.79167008097611,0 11.24826933207685,43.79230733776345,0 11.24608850202978,43.79212956957394,0 
> 					</coordinates>
> 				</LinearRing>
> 			</outerBoundaryIs>
> 		</Polygon>
> 	</Placemark>
> </Document>
> </kml>
> ```
> 
> Above is the kml extract from example.kmz. It get styleUrl as feature property

@volterra79 I'm not that much into the code, do you mean it's a bug of the current **Open Layers** version?

volterra79 commented:
@Raruto ol.format.KML is in charge to extract features with properties from kml file. I don't know if is a bug of OL v5.3.0( version that g3w-client use). I need to check

Raruto committed:
jsdoc + refactor function `MapService::addExternalLayer(externalLayer, options={})`

Raruto committed:
refactor functions: `QueryResultsService::_addVectorLayersDataToQueryResponse` and `QueryResultsService::getVectorLayerFeaturesFromQueryRequest(vectorLayer, query={})`

volterra79 committed:
Add comment

Raruto committed:
refactor `Catalog.vue` listeners

Raruto committed:
Merge branch 'v3.6_feature-add-external-layer-to-querybypolygon' of https://github.com/g3w-suite/g3w-client into v3.6_feature-add-external-layer-to-querybypolygon

volterra79 committed:
Add atlas dialog title translation

Raruto committed:
jsdoc

Raruto committed:
challenge accepted, let's make it unreadable... 🤪

Raruto committed:
refactor function `DataQueryService::coordinates()`

Raruto commented:
@volterra79 I'm trying again with that file: https://github.com/g3w-suite/g3w-client/pull/356#issuecomment-1441609067

Do you get the following error too? Or maybe I broke something else with in my latest commits?

![image](https://user-images.githubusercontent.com/9614886/221148010-10acacbc-5797-45bb-91de-7ed780e3f9e5.png)

volterra79 commented:
> @volterra79 I'm trying again with that file: [#356 (comment)](https://github.com/g3w-suite/g3w-client/pull/356#issuecomment-1441609067)
> 
> Do you get the following error too? Or maybe I broke something else with in my latest commits?
> 
> ![image](https://user-images.githubusercontent.com/9614886/221148010-10acacbc-5797-45bb-91de-7ed780e3f9e5.png)

@Raruto i think something is broke after last commit. I investigate on it

volterra79 committed:
:bug: Fix https://github.com/g3w-suite/g3w-client/pull/356#issuecomment-1443475920 due a wrong refactoring. Pay attention next time

volterra79 committed:
Change in explicit way

Raruto commented:
Nice catch: https://github.com/g3w-suite/g3w-client/pull/356/commits/c03030d716718084f05ecf3b6bed70c27e3bff6f (right now without a proper IDE support + error tracing it seems to me to be a bit blind monkey 🙈)

Same thing below, clicking on the polygon at that point throws the following error:

![image](https://user-images.githubusercontent.com/9614886/221193716-88ad0092-769f-4a93-b310-e9d40c94f7f8.png)

```js
TypeError: query.layer.getName is not a function
    info QueryResults.vue:239
    get vendor.min.js:20897
    evaluate vendor.min.js:20897
    Rt vendor.min.js:20897
    render QueryResults.vue:6
    _render vendor.min.js:20897
    r vendor.min.js:20897
    get vendor.min.js:20897
    run vendor.min.js:20897
    Et vendor.min.js:20897
    Ge vendor.min.js:20897
    Ke vendor.min.js:20897
    promise callback*Be vendor.min.js:20897
    Ge vendor.min.js:20897
    update vendor.min.js:20897
    update vendor.min.js:20897
    notify vendor.min.js:20897
    <anonymous> vendor.min.js:20897
    clearState queryresultsservice.js:881
    setQueryResponse queryresultsservice.js:107
    callSetter g3wobject.js:122
    next g3wobject.js:154
    next g3wobject.js:174
    setter g3wobject.js:180
    _callee2$ main.js:766
    Babel 9
    _callee4$ main.js:749
    Babel 7
    _callee$ data.js:36
    Babel 7
    _callee4$/< data.js:73
    _callee4$ data.js:72
    Babel 7
[vendor.min.js:20897:11628](http://localhost:8000/static/client/js/vendor.min.js)
    He vendor.min.js:20897
    Fe vendor.min.js:20897
    Pe vendor.min.js:20897
    _render vendor.min.js:20897
    r vendor.min.js:20897
    get vendor.min.js:20897
    run vendor.min.js:20897
    Et vendor.min.js:20897
    Ge vendor.min.js:20897
    Ke vendor.min.js:20897
    (Async: promise callback)
    Be vendor.min.js:20897
    Ge vendor.min.js:20897
    update vendor.min.js:20897
    update vendor.min.js:20897
    notify vendor.min.js:20897
    <anonymous> vendor.min.js:20897
    clearState queryresultsservice.js:881
    setQueryResponse queryresultsservice.js:107
    callSetter g3wobject.js:122
    next g3wobject.js:154
    next g3wobject.js:174
    setter g3wobject.js:180
    _callee2$ main.js:766
    Babel 9
    _callee4$ main.js:749
    Babel 7
    _callee$ data.js:36
    Babel 7
    _callee4$/< data.js:73
    _callee4$ data.js:72
    Babel 7

```

volterra79 commented:
@Raruto feel free to solve it. I can't replicate it

Raruto commented:
> @Raruto feel free to solve it. I can't replace it

Personal reminder: investigate further if this is just a Firefox related issue, that's why after every click on the map the following error is reported:

```
XML Parsing Error: syntax error
Location: http://localhost:8000/en/map/buildings/qdjango/10/
Line Number 1, Column 1:
```

and that error is throwed three times in the previous case of clicking on the kmz file.

volterra79 committed:
Merge branch 'dev' into v3.6_feature-add-external-layer-to-querybypolygon

volterra79 committed:
Adjust toggled event  and add change-spatial-method event

volterra79 committed:
Group same behavior on polygon and bbox

volterra79 committed:
Handle change spatial method on query by bbox and query by polygon map control

volterra79 committed:
Fix enable disable add layer button when close modal window dialog

volterra79 committed:
When checkbox is change, need to be set visible property of external layer (no projectLayer) to checked boolean value to trigger change to vue property

volterra79 committed:
Fix and improve enable disable map control in case of change visibility and selected property of layers

Raruto commented on src/services/data-query.js:
```diff
@@ -28,47 +28,63 @@ function QueryService(){
+    condition      = this.condition,
+    excludeLayers  = []
+  } = {}) {
+    const hasExternalLayers = this.getSelectedExternalLayers({ type: "vector" }).length;
```
double check, I'm not so sure if this is the correct check 

if we set as in the figure: `hasExternalLayers == false`:

![image](https://user-images.githubusercontent.com/9614886/221603300-7a4b953f-4a39-48ee-82d0-d8cd5f9106b3.png)

Then I guess this may also be the origin of the fatal exception shown here: https://github.com/g3w-suite/g3w-client/pull/356#issuecomment-1443704286

Raruto replied:
That error https://github.com/g3w-suite/g3w-client/pull/356#issuecomment-1443704286 is triggered at the following line because the `query.layer.getName()` function isn't added to object queries performed on external layers?

https://github.com/g3w-suite/g3w-client/blob/716fdffb9203542e5677e1529872df5ffa4fd8fc/src/components/QueryResults.vue#L239

Raruto committed:
update jsdoc style

Raruto committed:
mispelling

Raruto committed:
slim down callback

volterra79 committed:
Filter only visible external layer have to be add to result

Raruto committed:
spacing

volterra79 committed:
Merge branch 'dev' into v3.6_feature-add-external-layer-to-querybypolygon

volterra79 committed:
Add hasExternalLayerSelected base method

volterra79 committed:
remove extarlLayers from query:polygon data query method parameter and ad layer parameter for polygon layer.
Add external object property on query in data-query.js methods to handle externalLayer on queryresultsservice.js

volterra79 committed:
add selectedLayer property and setSelectedLayer method to handle setEnable map control

Raruto commented on src/app/gui/queryresults/queryresultsservice.js:
```diff
@@ -1256,14 +1258,31 @@ proto.getVectorLayerFeaturesFromQueryRequest = function(vectorLayer, query={}){
+      return;
+    }
+
+    // add vector layers to query response
+    this._vectorLayers
+      .forEach(layer => {
+        // always visible
+        if (layer.getVisible()) {
+          if (true === queryResponse.query.external.filter.SELECTED && !catalogService.isExternalLayerSelected({ id: layer.get('id'), type: 'vector' })) {
+            return;
+          } else if (false === queryResponse.query.external.filter.SELECTED && catalogService.isExternalLayerSelected({ id: layer.get('id'), type: 'vector'})) {
+            return;
+          } else {
+            queryResponse.data.push(this.getVectorLayerFeaturesFromQueryRequest(layer, queryResponse.query));
+          }
+        }
+      });
   });
```
here it can be a bit difficult to follow the logic of the code

there are many return conditions inside a loop whose intent is to push new elements into an array.

Try to check if it can be equivalent re-written like below:


```suggestion
    this._vectorLayers
      .forEach(layer => {
        const isExternalLayerSelected = catalogService.isExternalLayerSelected({ id: layer.get('id'), type: 'vector' });
        const isExternalLayerFilter   = queryResponse.query.external.filter.SELECTED;
        if (
          layer.getVisible() &&
          (false === isExternalLayerFilter || isExternalLayerSelected) &&
          (true === isExternalLayerFilter || !isExternalLayerSelected)
          ) {
            queryResponse.data.push(this.getVectorLayerFeaturesFromQueryRequest(layer, queryResponse.query));
        }
      });
  });
```

Raruto commented on src/services/data-query.js:
```diff
@@ -28,47 +28,78 @@ function QueryService(){
-    const polygonLayer = excludeLayers[0];
-    const fid = feature.get(G3W_FID);
+  this.polygon = function({
+    feature,
+    feature_count   = this.project.getQueryFeatureCount(),
+    filterConfig    = {},
+    multilayers     = false,
+    condition       = this.condition,
+    /**
+     * @since v3.8
+     */
+    layer    = null,
+    excludeSelected = null
```
New parameters have been added:

```js
layer           = null,
excludeSelected = null
```

and the following has been removed:

```js
excludeLayers = []
```

Is this intentional?

Double check if we need to write a fallback condition within the `polygon()` function in order to avoid any new potential unexpected errors.

volterra79 committed:
Changes accord to https://github.com/g3w-suite/g3w-client/pull/356#pullrequestreview-1319204178

Raruto commented on src/services/data-query.js:
```diff
@@ -81,48 +112,115 @@ function QueryService(){
+      external: {
+        add: true,
+        filter: {
+          SELECTED: externalSelectedLayers.length > 0
```
The `externalSelectedLayers` array could be undefined?

It seems to be undeclared within the `bbox()` scope.

volterra79 replied:
@Raruto solved https://github.com/g3w-suite/g3w-client/pull/356/commits/03c52cc14f2f8290ff54f0372e47ce87798776d8

volterra79 committed:
Fix changing variable refactoring not existing

Raruto commented on src/services/data-query.js:
```diff
@@ -81,48 +112,115 @@ function QueryService(){
-      coordinates,
-      type: 'coordinates',
-    });
+    /**
+     * @since v3.8 in case of project layer selected need to be set add external property to false to avoid add
+     * external layer to results
+     */
+    query.external.add = !(1 === layers.length && layers[0].isSelected());
```
If I understand correctly, you mean like this?

```suggestion
    // set external property `add: false` in case
    // of selected layer in order to avoid querying
    // a temporary layer (external layer)
    if (1 === layers.length && layers[0].isSelected()) {
      query.external.add = false;
    }
```

volterra79 replied:
@Raruto right

volterra79 committed:
Reset to default (initial) value of spatial Method

Raruto committed:
jsdoc

volterra79 committed:
Update src/services/data-query.js

Co-authored-by: Raruto <Raruto@users.noreply.github.com>

Raruto committed:
Merge branch 'v3.6_feature-add-external-layer-to-querybypolygon' of https://github.com/g3w-suite/g3w-client into v3.6_feature-add-external-layer-to-querybypolygon

volterra79 committed:
🔥 Remove this._changeMapMapControls.push(change) because it not more necessary. It was referred to v3.3 way of change project

volterra79 committed:
Merge branch 'v3.6_feature-add-external-layer-to-querybypolygon' of https://github.com/g3w-suite/g3w-client into v3.6_feature-add-external-layer-to-querybypolygon

Raruto committed:
rename variable: `isFilterSelected` into `isExternalFilterSelected`

Raruto committed:
Merge branch 'v3.6_feature-add-external-layer-to-querybypolygon' of https://github.com/g3w-suite/g3w-client into v3.6_feature-add-external-layer-to-querybypolygon

Raruto committed:
cuts of alias variables

perhaps harder to read, keep a focus on the return values

Raruto commented:
@volterra79 I'm getting different feature id values depending on the currently selected layer:

## `FID: 0` Query by Bolygon: "external layer selected"
![image](https://user-images.githubusercontent.com/9614886/222103707-ea165607-8c2a-41a3-b8fc-ff9ee35e7e69.png)

## `FID: undefined` Query by Polygon: "project layer selected"
![image](https://user-images.githubusercontent.com/9614886/222104308-aa75f5b3-a8e7-4bdd-a765-76c99861d3d9.png)

in both cases:

- I'm using [example.kmz.zip](https://github.com/g3w-suite/g3w-client/files/10813013/example.kmz.zip) as external layer
- I click on the point shown on the map (where there are no overlapping layers)

volterra79 commented:
@Raruto i check

volterra79 committed:
🐛 Fix missing layer during refactoring

volterra79 committed:
Remove console.log

volterra79 committed:
Move reset default spatial method when map control querybybbox and querybypolygon are not active

volterra79 committed:
Fix get toggled value from event. Previous get event and never be false

Raruto committed:
broken html

Raruto committed:
external link

Raruto committed:
code indentation

Raruto committed:
lazy deprecation

Raruto committed:
vertical spacing

Raruto committed:
stripped down comments

Raruto committed:
invalid return type

Raruto committed:
inline nested switch

Raruto committed:
code indentation

Raruto committed:
jsdoc

Raruto committed:
jsdoc + lazy deprecation

Raruto committed:
jsdoc parameters

Raruto commented on src/app/g3w-ol/controls/querybypolygoncontrol.js:
```diff
@@ -87,15 +87,16 @@ proto.listenLayersVisibilityChange = function() {
-    const {state} = layer;
     this.unwatches.push(
-      VM.$watch(() =>  state.visible, visible => {
+      VM.$watch(() =>  layer.state.visible, visible => {
```
double check if coded in this way can lead to additional issues

That means, before there was an object destructuring + function clousure for the `layer.state` property?

Raruto commented:
@volterra79 just to know, is it intended that you can't select / use a single category? (eg. for a `querybypolygon` request)

![image](https://user-images.githubusercontent.com/9614886/222464944-4fffc865-9d17-404a-9f8d-601171937dc8.png)

volterra79 commented:
> @volterra79 just to know, is it intended that you can't select / use a single category? (eg. for a `querybypolygon` request)
> 
> ![image](https://user-images.githubusercontent.com/9614886/222464944-4fffc865-9d17-404a-9f8d-601171937dc8.png)

@Raruto . At moment categories can be checked and unchecked. There isn't a request to select it. Maybe in the future

Raruto commented:
> At moment categories can be checked and unchecked. There isn't a request to select it. Maybe in the future

ok, just to figure out if it was a regression (since so much has changed here and fatigue is felt.. 🥴).
